media/heif/bmff/bmff.go

```diff
@@ -699,39 +699,163 @@ func parseItemLocationBox(outer *box, br *bufReader) (Box, error) {
 	if err != nil {
 		return nil, err
 	}
+	if fb.Version > 2 {
+		return nil, fmt.Errorf("go4 heif: error unsupported iloc version of %v", fb.Version)
+	}
 	ilb := &ItemLocationBox{
 		FullBox: fb,
 	}
-	buf, err := br.Peek(4)
+	buf, err := br.Peek(2)
 	if err != nil {
 		return nil, err
 	}
 	ilb.offsetSize = buf[0] >> 4
 	ilb.lengthSize = buf[0] & 15
 	ilb.baseOffsetSize = buf[1] >> 4
-	if fb.Version > 0 { // version 1
+	switch fb.Version {
+	case 0:
+		ilb.indexSize = 0
+	case 1, 2:
 		ilb.indexSize = buf[1] & 15
+	default:
+		return nil, fmt.Errorf("go4 heif: error unsupported iloc version of %v", fb.Version)
+	}
+	br.Discard(2)
+	switch fb.Version {
+	case 0, 1:
+		buf16, err := br.Peek(2)
+		if err != nil {
+			return nil, err
+		}
+		ilb.ItemCount = uint32(binary.BigEndian.Uint16(buf16))
+		br.Discard(2)
+	case 2:
+		buf32, err := br.Peek(4)
+		if err != nil {
+			return nil, err
+		}
+		ilb.ItemCount = binary.BigEndian.Uint32(buf32)
+		br.Discard(4)
+	default:
+		return nil, fmt.Errorf("go4 heif: error unsupported iloc version of %v", fb.Version)
 	}
-
-	ilb.ItemCount = binary.BigEndian.Uint16(buf[2:4])
-	br.Discard(4)
 
 	for i := 0; br.ok() && i < int(ilb.ItemCount); i++ {
 		var ent ItemLocationBoxEntry
-		ent.ItemID, _ = br.readUint16()
-		if fb.Version > 0 { // version 1
+		switch fb.Version {
+		case 0, 1:
+			itemId16, _ := br.readUint16()
+			ent.ItemID = uint32(itemId16)
+		case 2:
+			ent.ItemID, _ = br.readUint32()
+		default:
+			return nil, fmt.Errorf("go4 heif: error unsupported iloc version of %v", fb.Version)
+		}
+		switch fb.Version {
+		case 0:
+			// Do Nothing as this field doesn't exist in V0 iloc
+		case 1, 2:
 			cmeth, _ := br.readUint16()
 			ent.ConstructionMethod = byte(cmeth & 15)
+		default:
+			return nil, fmt.Errorf("go4 heif: error unsupported iloc version of %v", fb.Version)
 		}
 		ent.DataReferenceIndex, _ = br.readUint16()
 		if br.ok() && ilb.baseOffsetSize > 0 {
-			br.Discard(int(ilb.baseOffsetSize) / 8)
+			// Only 0, 4 and 8 are valid as at 2025-01-18.
+			switch ilb.baseOffsetSize {
+			case 0:
+				// Do Nothing as there are zero bytes to read
+			case 4:
+				buf32, err := br.Peek(4)
+				if err != nil {
+					return nil, err
+				}
+				ent.BaseOffset = uint64(binary.BigEndian.Uint32(buf32))
+				br.Discard(4)
+			case 8:
+				buf64, err := br.Peek(8)
+				if err != nil {
+					return nil, err
+				}
+				ent.BaseOffset = binary.BigEndian.Uint64(buf64)
+				br.Discard(8)
+			default:
+				return nil, fmt.Errorf("go4 heif: error baseOffsetSize of %v is not supported", ilb.baseOffsetSize)
```

Comment:
One should prefer adding context to error messages towards the end of the error message. This maximizes the contiguous text that can be used to grep or search for the code a specific log line of errors comes from.
```suggestion
				return nil, fmt.Errorf("baseOffsetSize not supported: %d", ilb.baseOffsetSize)
```

We would probably also want to use whatever name in the error message that the specification uses? (Principle: prefer error messages that are useful to users, not the developers of this package.) Though, I cannot find an appropriate standard from a quick web search.

Reply:
There didn't seem to be any consistency in the existing error messages.
I've applied your requests.